src/ML/src/MANNAutoregressive.cpp

```diff
@@ -0,0 +1,706 @@
+                                input.desiredFutureFacingDirections,
+                                1.3,
+                                m_pimpl->mannInput.facingDirectionTrajectory.rightCols<6>());
+
+    m_pimpl->trajectoryBlending(mannOutput.futureBaseVelocitiesTrajectory,
+                                input.desiredFutureBaseVelocities,
+                                1.3,
+                                m_pimpl->mannInput.baseVelocitiesTrajectory.rightCols<6>());
+
+    if (!m_pimpl->mann.setInput(m_pimpl->mannInput))
+    {
+        log()->error("{} Unable to set the input to MANN network", logPrefix);
+        return false;
+    }
+
+    constexpr double toleranceRobotStopped = 0.05;
+    m_pimpl->isRobotStopped
+        = m_pimpl->mannInput.basePositionTrajectory
+              .isApprox(m_pimpl->state.previousMannInput.basePositionTrajectory,
+                        toleranceRobotStopped)
+          && m_pimpl->mannInput.baseVelocitiesTrajectory
+                 .isApprox(m_pimpl->state.previousMannInput.baseVelocitiesTrajectory,
+                           toleranceRobotStopped)
+          && m_pimpl->mannInput.facingDirectionTrajectory
+                 .isApprox(m_pimpl->state.previousMannInput.facingDirectionTrajectory,
+                           toleranceRobotStopped)
+          && m_pimpl->mannInput.jointVelocities
+                 .isApprox(m_pimpl->state.previousMannInput.jointVelocities, toleranceRobotStopped)
+          && m_pimpl->mannInput.jointPositions
+                 .isApprox(m_pimpl->state.previousMannInput.jointPositions, toleranceRobotStopped);
+
+    m_pimpl->state.I_H_FD = I_H_current_FD;
+    m_pimpl->state.previousMannInput = m_pimpl->mannInput;
+
+    return true;
+}
+
+void MANNAutoregressive::Impl::trajectoryBlending(
+    Eigen::Ref<const Eigen::Matrix2Xd> mannOutputMatrix,
+    Eigen::Ref<const Eigen::Matrix2Xd> desiredMatrix,
+    const double& tau,
+    Eigen::Ref<Eigen::Matrix2Xd> out)
+{
+    for (int i = 0; i < mannOutputMatrix.cols(); i++)
+    {
+        const double T = std::pow((i + 1.0) / double(mannOutputMatrix.cols()), tau);
+        out.col(i) = (1 - T) * mannOutputMatrix.col(i) + T * desiredMatrix.col(i);
+    }
+}
+
+bool MANNAutoregressive::advance()
+{
+    using namespace std::chrono_literals;
+    using namespace BipedalLocomotion::GenericContainer::literals;
+
+    constexpr auto logPrefix = "[MANNAutoregressive::advance]";
+
+    // invalidate the output
+    m_pimpl->isOutputValid = false;
+
+    // check the status of the state machine.
+    if (m_pimpl->fsmState != Impl::FSM::Reset && m_pimpl->fsmState != Impl::FSM::Running)
+    {
+        log()->error("{} You need to reset the network before calling advance the first time. "
+                     "Please call reset().",
+                     logPrefix);
+        return false;
+    }
+
+    // perform one iteration of MANN
+    if (!m_pimpl->mann.advance() || !m_pimpl->mann.isOutputValid())
```

Comment:
What if I forgot to call ``initialize``?

Reply:
I updated the state machine. Now it is impossible to call reset if the advanceable has not been initialized. So in this case advance it will not run since the MANNAutoregressive will not be in the reset state.